parser.go

```diff
@@ -73,3 +76,61 @@ func SelectorSpecFromPath(path Expression, optionalSubselectorAtTarget builder.S
 
 	return ss, nil
 }
+
+type selectorSpec struct {
+	n datamodel.Node
+	s selector.Selector
+}
+
+func (ss selectorSpec) Node() datamodel.Node {
+	return ss.n
+}
+
+func (ss selectorSpec) Selector() (selector.Selector, error) {
+	return ss.s, nil
+}
+
+/*
+SelectorSpecFromJson transforms a string form of a dag-json representation of a
+go-ipld-prime selector into a selector-spec object.
+
+The full dag-json form of a selector is not a user-friendly thing to construct,
+however it does allow for the representation of the full range of possible
+selectors where simple pathing-based selectors won't work.
+
+Note that using this format, the top-level of the input must be wrapped in an
+envelope (map) with a single field "selector".
+
+For example, a selector for the path x/y/10/z could be supplied to
+SelectorSpecFromJson as:
+
+	{"selector"{"f":{"f>":{"x":{"f":{"f>":{"y":{"f":{"f>":{"10":{"f":{"f>":{"z":{".":{}}}}}}}}}}}}}}}
+
+See https://ipld.io/specs/selectors/ for more information.
+*/
+func SelectorSpecFromJson(jsonStr string) (builder.SelectorSpec, error) {
```

Comment:
Great stuff!
My only concern here is the asymmetry of both calls:

`func SelectorSpecFromPath(path Expression, optionalSubselectorAtTarget builder.SelectorSpec)`

vs

`func SelectorSpecFromJson(jsonStr string) (builder.SelectorSpec, error)`

Not sure if the machinery is even compatible here though... so perhaps worthwhile to add everything as-is

Reply:
yeah, I had it the same but then realised I had no idea how to munge the two things together! It'd force you to figure out if the selector had a logical place to insert a subselector first and then do the stitching, which seems 😬 so best to leave it off?

Reply:
:shipit:

Reply:
@rvagg I was thinking more about this - there is something missing conceptually ( not the code, selector-space in general ).

In general selectors are never standalone: they require a "starting point" for the traversal. Also in general this starting-node is **not** described within the selector itself. Instead when you consume a selector you bring two pieces of knowledge to the table: *where to start*, and *what to do*

The way the "path" part of this library is used in filecoin today, is:
- [Tack on a recursor to the path](https://github.com/filecoin-project/lotus/blob/v1.13.0-rc1/node/impl/client/client.go#L845-L867), grab the entire thing given the start describe in the deal
- [Re-evaluate the selector over what we grabbed, find the new root](https://github.com/filecoin-project/lotus/blob/v1.13.0-rc1/node/impl/client/client.go#L1028-L1068)
- [Give the new start + implicit recursor to the unixfs engine](https://github.com/filecoin-project/lotus/blob/v1.13.0-rc1/node/impl/client/client.go#L1070-L1075)
- Profit

The above workflow becomes impossible with a selector as expressed in this PR: there is no way to communicate what the "new start" would be, so you can not reify the result ( unixfs or whatever it is ). We touched on this [higher up](https://github.com/ipld/go-ipld-selector-text-lite/pull/4#discussion_r707056267), but I didn't click on the implications, so jolting them down.

What this code *can* be used for is doing [a retrieval only, without reification](https://github.com/filecoin-project/lotus/blob/v1.13.0-rc1/node/impl/client/client.go#L971-L976). This implies the user assembling a complete selector using your new code, and then assembling a new "start-point" path selector and applying to that to what ended up in their blockstore. Which... unergonomic, and a nightmare to put behind an API surface...

Ideally a selector, to be useful, ought to also be able to specify the entire "playbook" within itself...
@rvagg @willscott @warpfork what are your thoughts?

Reply:
don't we want a starting point CID, and a selector from that CID?

Reply:
@willscott ideally yes, today we don't have that. What we have is:
- Starting point of complete dag ( root cid in `label` )
- Path to the starting point of the sub-dag within that dag ( the selector-lite-path thing )
- Implicit need to reify the subdag as a unixfs "thing"

Reply:
@willscott another way to phrase this is "we lack ability to specify conditional recursion"

Reply:
we can do this potentially do this in 2 iterations, right? first using a selector (or the selector-lite-path expression into a selector) to get the CID of the actual starting point.
Then a second request based on the actual starting point, and we'd hopefully have hamt-signalling-in-selectors to better select over unixfs. There's a MVP version of that signalling that should be able to be introduced without too much pain.

Reply:
@willscott you certainly could, though if this is a paid retrieval doing the deed twice is... rough UX
In any case - until non-piece-root-cid retrieval works in lotus, this is all moot: what we merged was the only thing possible at the time.

Reply:
for miners with the dagstore, the non-piece-root-cid retrieval is already working, right?